### DISCUSSION TITLE
Trying to understand what Schrodinger is/isn't does/doesn't.

### DISCUSSION THREAD

**Discussion by dragonfly-ai** [armanbilge/schrodinger#313] [General]
Is this an attempt to make Scala a viable [Probabilistic Programming Language(PPL)](https://en.wikipedia.org/wiki/Probabilistic_programming#List_of_probabilistic_programming_languages)?

How does it diverge from [rainier?](https://github.com/stripe/rainier) or [Figaro](https://github.com/charles-river-analytics/figaro)?

Documentation questions:
1.  Do you have plans to create example programs to help articulate what Schrodinger does and how to use it?
2.  "At its heart is the RVT monad transformer for modeling pseudo-random effects that is designed to fully compose with asynchronous effects such as IO."  Doesn't this mean that we can use Schrodinger to make generative models?  In other words, to generate pseudo-random numbers from various kinds of probability distributions?  If that's not what you mean, what, then?
3.  "The goal is that this construct will facilitate the implementation of high-performance, concurrent Monte Carlo algorithms for simulation and inference."  Simulation sounds like generative models.  Inference sounds like estimation.  Does Schrodinger support generating random events from a specified probability distribution, or does it support estimating the parameters of a specific probability distribution from a set or stream of random events?  Or does it do both?  Neither?  For example, to use Schrodinger do I say: here's a Beta distribution defined by these parameters, please generate some random values from that distribution.  Or do I say:  Here's a bunch of data, please fit a Beta distribution to it.  Or do I say:  Here's a bunch of data, please choose a good probability distribution then estimate its parameters?

Whichever of these things Schrodinger can do, the README.md could really use some examples.

**Comment by armanbilge** [comment]
> Is this an attempt to make Scala a viable [Probabilistic Programming Language(PPL)](https://en.wikipedia.org/wiki/Probabilistic_programming#List_of_probabilistic_programming_languages)?

Ha, that makes it sound way more ambitious. I would describe is an experimental attempt at embedding a probabilistic programming language in Scala.

---

> How does it diverge from [rainier?](https://github.com/stripe/rainier)

Rainier is not a general probabilistic programming language: it is restricted to models which can be fit using Hamiltonian Monte Carlo. Also, that is pretty much all you can do with the models: fit them with HMC.

Schrodinger allows you to express arbitrary probabilistic programs. It also does not restrict what you do with them: you can simulate from them, fit them with arbitrary Monte Carlo algorithms, or apply various transformations.

Indeed, I believe it should be possible to implement a Rainier interpreter for Schrodinger, such that probabilistic programs expressed using Schrodinger's abstractions could be fit with Rainier.

> or [Figaro](https://github.com/charles-river-analytics/figaro)?

Figaro looks more similar to Schrodinger. However, Schrodinger follows the style of Typelevel / Cats libraries and accordingly takes more cues from category theory. For example, Figaro appears to have introduced an ad-hoc concept of `Chain` which is actually a well-known concept of `flatMap` operation on a `Monad`.

---

> Do you have plans to create example programs to help articulate what Schrodinger does and how to use it?

Concrete plans? Not really, no :) vague plans of future documentation? Sure, definitely!

Truthfully, I'm still figuring out what Schrodinger does and how to use it 😁 

---

> If that's not what you mean, what, then?

The key phrase there is "fully compose with asynchronous effects such as `IO`". I want a probabilistic programming language with first-class support for parallelism and asynchronous I/O i.e. networking. `IO` enables both of those.

- I want parallelism so that I can distribute intensive calculations across multiple CPUs.
- I want asynchronous I/O so that my programs can communicate across multiple machines over the network and read/write results from databases

In summary, I want to be able to deploy a probabilistic program as a distributed system.

---

> Simulation sounds like generative models.  Inference sounds like estimation.

Bayesian inference is about simulating from a posterior distribution to do estimation. Schrodinger supports simulation, therefore it supports estimation via Bayesian inference.

---

> Whichever of these things Schrodinger can do, the README.md could really use some examples.

The tl;dr is that Schrodinger is an experimental library whose design goals are:

1. to facilitate implementation of distributed systems for Bayesian inference

2. to follow Typelevel design principles and best practices

Both of these are complex topics that will require exposition.

**Comment by dragonfly-ai** [comment]
> 1. implementation of distributed systems for Bayesian inference

That sounds really cool.

Thanks for the elaborations.  They were helpful.

I guess I'll have to learn Cats and more about Probabilistic Programs before I can understand any better.
